consensus/spos/scheduledProcessor.go

```diff
@@ -0,0 +1,180 @@
+package spos
+
+import (
+	"sync"
+	"time"
+
+	"github.com/ElrondNetwork/elrond-go/core/check"
+	"github.com/ElrondNetwork/elrond-go/data"
+	"github.com/ElrondNetwork/elrond-go/ntp"
+	"github.com/ElrondNetwork/elrond-go/process"
+)
+
+type processingStatus int
+
+const (
+	processingNotStarted processingStatus = 0
+	processingError      processingStatus = 1
+	inProgress           processingStatus = 2
+	processingOK         processingStatus = 3
+
+	processingNotStartedString = "processing not started"
+	processingErrorString      = "processing error"
+	inProgressString           = "processing in progress"
+	processingOKString         = "processing OK"
+	unexpectedString           = "processing OK"
+
+	processingCheckStep = 10 * time.Millisecond
+)
+
+// String returns human readable processing status
+func (ps processingStatus) String() string {
+	switch ps {
+	case processingNotStarted:
+		return processingNotStartedString
+	case processingError:
+		return processingErrorString
+	case inProgress:
+		return inProgressString
+	case processingOK:
+		return processingOKString
+	default:
+		return unexpectedString
+	}
+}
+
+// ScheduledProcessorArgs holds the arguments required to instantiate the pipelineExecution
+type ScheduledProcessorArgs struct {
+	SyncTimer                  ntp.SyncTimer
+	Processor                  process.ScheduledBlockProcessor
+	ProcessingTimeMilliSeconds uint32
+}
+
+type scheduledProcessor struct {
+	syncTimer      ntp.SyncTimer
+	processingTime time.Duration
+	processor      process.ScheduledBlockProcessor
+
+	startTime time.Time
+	status    processingStatus
+	sync.RWMutex
+
+	oneInstance sync.RWMutex
+}
+
+// NewScheduledProcessor creates a new processor for scheduled transactions
+func NewScheduledProcessor(args ScheduledProcessorArgs) (*scheduledProcessor, error) {
+	if check.IfNil(args.SyncTimer) {
+		return nil, process.ErrNilSyncTimer
+	}
+	if check.IfNil(args.Processor) {
+		return nil, process.ErrNilBlockProcessor
+	}
+
+	return &scheduledProcessor{
+		syncTimer:      args.SyncTimer,
+		processingTime: time.Duration(args.ProcessingTimeMilliSeconds) * time.Millisecond,
+		processor:      args.Processor,
+		status:         processingNotStarted,
+	}, nil
+}
+
+func (sp *scheduledProcessor) computeRemainingProcessingTime() time.Duration {
+	currTime := sp.syncTimer.CurrentTime()
+
+	sp.RLock()
+	elapsedTime := currTime.Sub(sp.startTime)
+	sp.RUnlock()
+
+	if elapsedTime < 0 {
+		return 0
+	}
+
+	return sp.processingTime - elapsedTime
+}
+
+// IsProcessedOK returns true if the scheduled processing was finalized without error
+// Function is blocking until the allotted time for processing is finished
+func (sp *scheduledProcessor) IsProcessedOK() bool {
+	status := sp.getStatus()
+
+loop:
```

Comment:
the break loop stuff could have been avoided by extracting the for loop in a new function and used return statement.

Reply:
:+1: